### PR TITLE
Migrate Linux builds to GitHub Actions

### PR DESCRIPTION
**Pull request by Saviq**
---
Remaining in Travis:
- macOS builds
- daily snap processing

Missing:
- build comments (need to decide what to do there…)

### CODE REVIEW

File: .github/actions/build-params/action.yml [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        echo "::set-output name=channel::edge/pr${BASH_REMATCH[1]}"
+
+      # If it's a release tag or branch, publish to the `candidate` channel.
+      elif [[ ${{ github.ref }} =~ ${MULTIPASS_RELEASE_BRANCH_PATTERN} \
+        || ${{ github.ref }} =~ ${MULTIPASS_RELEASE_TAG_PATTERN} ]]; then
+        echo "::set-output name=channel::candidate"
+
+      # All other pushes annotate with the CI run number.
+      elif [ ${{ github.event_name }} == "push" ]; then
+        echo "::set-output name=label::ci${{ github.run_number }}"
```
Comment (Saviq) [inline .github/actions/build-params/action.yml:35]: This can go away.
```suggestion
```

File: .github/actions/checkout/action.yml [hunk 2]
```diff
@@ -0,0 +1,20 @@
+name: Checkout
+description: Check out Multipass code along with tags and submodules
+
+runs:
+  using: composite
+  steps:
+  - name: Check out submodules
+    shell: bash
+    working-directory: ${{ github.workspace }}
```
Comment (Saviq) [inline .github/actions/checkout/action.yml:9]:
```suggestion
```

File: .github/actions/checkout/action.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    run: |
+      git submodule sync
+      # This repository can't do depth.
+      git -c protocol.version=2 submodule update --init 3rd-party/xz-decoder/xz-embedded
+      git -c protocol.version=2 submodule update --init --force --depth=1 --recursive
+
+  - name: Fetch tags
+    shell: bash
+    working-directory: ${{ github.workspace }}
```
Comment (Saviq) [inline .github/actions/checkout/action.yml:18]:
```suggestion
```

File: .github/workflows/linux.yml [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      uses: ./.github/actions/checkout
+
+    - name: Determine build parameters
+      id: build-params
+      uses: ./.github/actions/build-params
+
+    - name: Patch
+      run: |
+        [ ! -f tests/travis.patch ] || patch -p1 --no-backup-if-mismatch < tests/travis.patch
+        [ ! -f tests/travis-${{ matrix.build-type }}.patch ] || patch -p1 --no-backup-if-mismatch < tests/travis-${{ matrix.build-type }}.patch
```
Comment (Saviq) [inline .github/workflows/linux.yml:80]: Maybe these can go somewhere else.

File: .github/workflows/linux.yml [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        key: ccache-${{ runner.os }}-${{ matrix.build-type }}-${{ steps.setup-ccache.outputs.key }}
+        restore-keys: |
+          ccache-${{ runner.os }}-${{ matrix.build-type }}-
+        path: ~/.ccache/**
+
+    - name: Set up coverage
+      if: ${{ matrix.build-type == 'Coverage' }}
+      run: |
+        mkdir "${HOME}/parts"
+        sg lxd -c '/snap/bin/lxc profile device add default build disk source=${HOME}/parts path=/root/parts'
```
Comment (Saviq) [inline .github/workflows/linux.yml:105]:
```suggestion
        sg lxd -c '/snap/bin/lxc profile device add default build disk source=${HOME}/parts/multipass/build path=/root/parts/multipass/build'
```

File: .github/workflows/linux.yml [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    - name: Build
+      run: |
+        # Inject the build label.
+        if [ -n "${{ steps.build-params.outputs.label }}" ]; then
+            sed -i "/configflags:/a \    - -DMULTIPASS_BUILD_LABEL=${{ steps.build-params.outputs.label }}" snap/snapcraft.yaml
+        fi
```
Comment (Saviq) [inline .github/workflows/linux.yml:112]:
```suggestion
        sed -i "/configflags:/a \    - -DMULTIPASS_BUILD_LABEL=${{ steps.build-params.outputs.label }}" snap/snapcraft.yaml
```

File: .github/workflows/linux.yml [hunk 7]
```diff
@@ -0,0 +1,198 @@
...
+
+    - name: Measure coverage
+      if: ${{ matrix.build-type == 'Coverage' }}
+      timeout-minutes: 5
+      run: |
+        sg lxd -c '/snap/bin/lxc start snapcraft-multipass'
+        sg lxd -c \
+          '/snap/bin/lxc exec snapcraft-multipass -- \
+             cmake --build /root/parts/multipass/build --target covreport'
+        bash <(curl -s https://codecov.io/bash) -Z -s ${HOME}/parts
```
Comment (Saviq) [inline .github/workflows/linux.yml:139]:
```suggestion
        bash <(curl -s https://codecov.io/bash) -Z -s ${HOME}/parts/multipass/build
```

Comment (Saviq) [reply .github/workflows/linux.yml:139]: Maybe annotate? Retry?

File: .github/workflows/linux.yml [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    - name: Publish the snap
+      run: |
+        snapcraft upload *.snap --release ${{ needs.Build-Test.outputs.channel }}
+
+  # Report result to Bors on `staging` and `trying` branches.
+  Bors:
+    if: ${{ always() && github.event_name == 'push' && contains('refs/heads/staging refs/heads/trying', github.ref) }}
+    needs:
+    - Lint
+    - Build-Test
```
Comment (Saviq) [inline .github/workflows/linux.yml:191]: Verify these

File: .github/workflows/linux.yml [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    steps:
+    - name: Check out code
+      uses: actions/checkout@v2
+      with:
+        fetch-depth: 2
+
+    - name: Run clang-format through the diff
+      uses: ./.github/actions/lint
+
+  Build-Test:
```
Comment (townsend2010) [inline .github/workflows/linux.yml:29]: This was kind of confusing to me, ie, it seems it's building something called "Test".  I don't think spaces are allowed, so maybe something called `BuildAndTest` or `Build-and-Test` or something.  I think having an "and" in there makes it more clear that it's building *and* testing.

Comment (Saviq) [reply .github/workflows/linux.yml]: Done.

File: .github/workflows/linux.yml (hunk 4 above)
Comment (Saviq) [reply .github/workflows/linux.yml]: Moved to `.github/workflows/linux-*`.

File: .github/workflows/linux.yml (hunk 7 above)
Comment (Saviq) [reply .github/workflows/linux.yml]: Filed #1709.

**Comment by Saviq** [comment]
> Other than my comment about `Build-Test`, this seems good. Of course, there are a couple of little things left from the pair session yesterday, but they are not blockers.

All done, I believe.

**Pull request merged**